File: C11/백준/Silver/2108. 통계학/통계학.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int round(double x) {
    if (x >= 0){
		return (int)(x + 0.5);	
	}
    else {
		return (int)(x - 0.5);	
	}
}

int Mode(int N, int arr[]) {
    int i;
    int min = arr[0], max = arr[0];

    for (i = 1; i < N; i++) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }

    int range = max - min + 1;
    int* count = (int*)calloc(range, sizeof(int));

    for (i = 0; i < N; i++) {
        count[arr[i] - min]++;
    }

    int maxCount = 0;
    for (i = 0; i < range; i++) {
        if (count[i] > maxCount) {
            maxCount = count[i];
        }
    }

    int* modes = (int*)malloc(N * sizeof(int));
    int modeCount = 0;
    for (i = 0; i < range; i++) {
        if (count[i] == maxCount) {
            modes[modeCount++] = i + min;
        }
    }

    qsort(modes, modeCount, sizeof(int), compare);
    int result = (modeCount >= 2) ? modes[1] : modes[0];

    free(count);
    free(modes);

    return result;
}


int main(void){
	
	int N;
	scanf("%d", &N);
	
	int sum = 0;
	int arr[N];
	for(int i = 0; i < N; i++){
		scanf("%d", &arr[i]);
		sum += arr[i];
	}
	
	qsort(arr, N, sizeof(int), compare);
	
	printf("%d\n", round((double) sum / N));
	printf("%d\n", arr[N / 2]);
	printf("%d\n", Mode(N, arr));
	printf("%d", arr[N - 1] - arr[0]);
	
	return 0;
}

File: Java/백준/Silver/2108. 통계학/통계학.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine()); //수의 개수
        int sum = 0;

        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            sum += arr[i];
        }

        Arrays.sort(arr);

        sb.append(Math.round((double) sum / N)).append("\n"); //산술 평균
        sb.append(arr[(N / 2)]).append("\n"); //중앙값
        sb.append(Mode(N, arr)).append("\n"); //최빈값
        sb.append(arr[N - 1] - arr[0]); //범위

        System.out.print(sb);
    }

    static int Mode(int N, int[] arr) {
        int min = Arrays.stream(arr).min().getAsInt();
        int max = Arrays.stream(arr).max().getAsInt();

        int[] count = new int[max - min + 1];

        for (int i = 0; i < N; i++) {
            count[arr[i] - min]++;
        }

        int maxCount = 0;
        for (int c : count) {
            if (c > maxCount) {
                maxCount = c;
            }
        }

        List<Integer> modes = new ArrayList<>();
        for (int i = 0; i < count.length; i++) {
            if (count[i] == maxCount) {
                modes.add(i + min);
            }
        }

        Collections.sort(modes);

        if (modes.size() >= 2) {
            return modes.get(1); //두 번째로 작은 최빈값
        } else {
            return modes.get(0); //가장 작은 최빈값
        }
    }
}